lib/client-options.js

```diff
@@ -129,6 +129,61 @@ function extend(baseOptions, userOptions) {
 }
 
 /**
+ * Creates a new instance of query options with the values from the user.
+ * When some values are not defined, it takes the default values from
+ * - {@link ExecutionProfile}.
+ * - {@link QueryOptions} from the default options.
+ * @param {Client} client
+ * @param {QueryOptions|function} userOptions
+ * @param {Function} [rowCallback]
+ * @param {Boolean} [logged]
+ */
+function createQueryOptions(client, userOptions, rowCallback, logged) {
+  var profile =
+    client.profileManager.getProfile(userOptions && userOptions.executionProfile) || utils.emptyObject;
+  // userOptions can be undefined and could be of type function (is an optional parameter)
+  userOptions = (!userOptions || typeof userOptions === 'function') ? utils.emptyObject : userOptions;
+  var defaultQueryOptions = client.options.queryOptions;
+  // Using fixed property names is 2 order of magnitude faster than dynamically shallow clone objects
```

Comment:
The behavior of things when it comes to performance with js always surprises me :D